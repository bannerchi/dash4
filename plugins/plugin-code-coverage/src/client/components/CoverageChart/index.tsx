/** @jsx jsx */
import { css, jsx } from '@emotion/core';

import { Pie, PieChart } from 'recharts';

const colors = {
	success: '#28a745',
	warning: '#ffc107',
	danger: '#dc3545',
};

interface ICoverageChartProps {
	title: string;
	coverage: number;
	threshold: [number, number];
	className?: string;
	dark?: boolean;
}

function getColorFromPercentage(percentage: number, threshold: [number, number]) {
	if (percentage < threshold[0]) {
		return colors.danger;
	}

	if (percentage < threshold[1]) {
		return colors.warning;
	}

	return colors.success;
}

export const CoverageChart = ({ title, coverage, threshold, className = '', dark = false }: ICoverageChartProps) => {
	return (
		<div
			className={className}
			css={css`
				position: relative;
				width: 120px;
				height: 120px;
			`}
		>
			<div
				css={css`
					position: absolute;
					top: 10px;
					left: 20px;
					width: 80px;
				`}
			>
				<PieChart width={80} height={80}>
					<Pie
						data={[{ value: coverage }, { value: 100 - coverage, fill: dark ? '#000' : '#ccc' }]}
						dataKey="value"
						cx="50%"
						cy="50%"
						innerRadius={25}
						outerRadius={35}
						fill={getColorFromPercentage(coverage, threshold)}
					/>
				</PieChart>
			</div>
			<div
				className={`h3`}
				css={css`
					position: absolute;
					top: 10px;
					left: 20px;
					height: 80px;
					width: 80px;
					line-height: 80px;
					text-align: center;
					font-size: 14px;
					font-weight: ${dark ? 400 : undefined};
				`}
			>
				{coverage}%
			</div>
			<div
				className={`h6`}
				css={css`
					position: absolute;
					bottom: 0;
					width: 100%;
					text-align: center;
					padding: 5px 15px;
					margin: 0;
					font-weight: ${dark ? 100 : undefined};
				`}
			>
				{title}
			</div>
		</div>
	);
};

export default CoverageChart;
