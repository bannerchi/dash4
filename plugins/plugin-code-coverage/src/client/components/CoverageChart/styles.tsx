interface IProps {
	dark?: boolean;
}

export const styles = {
	root: {
		position: 'relative',
		width: 120,
		height: 120,
	},
	title: ({ dark = false }: IProps) => ({
		position: 'absolute',
		bottom: 0,
		width: '100%',
		textAlign: 'center',
		padding: '5px 15px',
		margin: 0,
		fontWeight: dark ? 100 : undefined,
	}),
	chart: {
		position: 'absolute',
		top: 10,
		left: 20,
		width: 80,
	},
	number: ({ dark = false }: IProps) => ({
		position: 'absolute',
		top: 10,
		left: 20,
		height: 80,
		width: 80,
		lineHeight: '80px',
		textAlign: 'center',
		fontSize: 14,
		fontWeight: dark ? 400 : undefined,
	}),
};
