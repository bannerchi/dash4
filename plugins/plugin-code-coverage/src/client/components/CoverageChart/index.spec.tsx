/// <reference types="@types/jest" />
import { render } from '@testing-library/react';
import React from 'react';
import 'react-dom';
import CoverageChart from './index';

describe('CoverageChart', () => {
	test('coverage danger', () => {
		const rendered = render(<CoverageChart title="title" coverage={10} threshold={[20, 50]} />);
		expect(rendered.container.firstChild).toMatchSnapshot();
	});

	test('coverage warning', () => {
		const rendered = render(<CoverageChart title="title" coverage={30} threshold={[20, 50]} />);
		expect(rendered.container.firstChild).toMatchSnapshot();
	});

	test('coverage success', () => {
		const rendered = render(<CoverageChart title="title" coverage={50} threshold={[20, 50]} />);
		expect(rendered.container.firstChild).toMatchSnapshot();
	});

	test('dark', () => {
		const rendered = render(<CoverageChart dark title="title" coverage={50} threshold={[20, 50]} />);
		expect(rendered.container.firstChild).toMatchSnapshot();
	});
});
