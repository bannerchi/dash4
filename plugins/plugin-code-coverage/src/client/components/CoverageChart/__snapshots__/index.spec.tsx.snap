// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CoverageChart coverage danger 1`] = `
.emotion-3 {
  position: relative;
  width: 120px;
  height: 120px;
}

.emotion-0 {
  position: absolute;
  top: 10px;
  left: 20px;
  width: 80px;
}

.emotion-1 {
  position: absolute;
  top: 10px;
  left: 20px;
  height: 80px;
  width: 80px;
  line-height: 80px;
  text-align: center;
  font-size: 14px;
}

.emotion-2 {
  position: absolute;
  bottom: 0;
  width: 100%;
  text-align: center;
  padding: 5px 15px;
  margin: 0;
}

<div
  class=" emotion-3"
>
  <div
    class="emotion-0"
  >
    <div
      class="recharts-wrapper"
      style="position: relative; cursor: default; width: 80px; height: 80px;"
    >
      <svg
        class="recharts-surface"
        height="80"
        version="1.1"
        viewBox="0 0 80 80"
        width="80"
      >
        <defs>
          <clippath
            id="recharts1-clip"
          >
            <rect
              height="70"
              width="70"
              x="5"
              y="5"
            />
          </clippath>
        </defs>
        <g
          class="recharts-layer recharts-pie"
        >
          <g
            class="recharts-layer"
          >
            <g
              class="recharts-layer recharts-pie-sector"
            />
            <g
              class="recharts-layer recharts-pie-sector"
            />
          </g>
        </g>
      </svg>
    </div>
  </div>
  <div
    class="h3 emotion-1"
  >
    10
    %
  </div>
  <div
    class="h6 emotion-2"
  >
    title
  </div>
</div>
`;

exports[`CoverageChart coverage success 1`] = `
.emotion-3 {
  position: relative;
  width: 120px;
  height: 120px;
}

.emotion-0 {
  position: absolute;
  top: 10px;
  left: 20px;
  width: 80px;
}

.emotion-1 {
  position: absolute;
  top: 10px;
  left: 20px;
  height: 80px;
  width: 80px;
  line-height: 80px;
  text-align: center;
  font-size: 14px;
}

.emotion-2 {
  position: absolute;
  bottom: 0;
  width: 100%;
  text-align: center;
  padding: 5px 15px;
  margin: 0;
}

<div
  class=" emotion-3"
>
  <div
    class="emotion-0"
  >
    <div
      class="recharts-wrapper"
      style="position: relative; cursor: default; width: 80px; height: 80px;"
    >
      <svg
        class="recharts-surface"
        height="80"
        version="1.1"
        viewBox="0 0 80 80"
        width="80"
      >
        <defs>
          <clippath
            id="recharts5-clip"
          >
            <rect
              height="70"
              width="70"
              x="5"
              y="5"
            />
          </clippath>
        </defs>
        <g
          class="recharts-layer recharts-pie"
        >
          <g
            class="recharts-layer"
          >
            <g
              class="recharts-layer recharts-pie-sector"
            />
            <g
              class="recharts-layer recharts-pie-sector"
            />
          </g>
        </g>
      </svg>
    </div>
  </div>
  <div
    class="h3 emotion-1"
  >
    50
    %
  </div>
  <div
    class="h6 emotion-2"
  >
    title
  </div>
</div>
`;

exports[`CoverageChart coverage warning 1`] = `
.emotion-3 {
  position: relative;
  width: 120px;
  height: 120px;
}

.emotion-0 {
  position: absolute;
  top: 10px;
  left: 20px;
  width: 80px;
}

.emotion-1 {
  position: absolute;
  top: 10px;
  left: 20px;
  height: 80px;
  width: 80px;
  line-height: 80px;
  text-align: center;
  font-size: 14px;
}

.emotion-2 {
  position: absolute;
  bottom: 0;
  width: 100%;
  text-align: center;
  padding: 5px 15px;
  margin: 0;
}

<div
  class=" emotion-3"
>
  <div
    class="emotion-0"
  >
    <div
      class="recharts-wrapper"
      style="position: relative; cursor: default; width: 80px; height: 80px;"
    >
      <svg
        class="recharts-surface"
        height="80"
        version="1.1"
        viewBox="0 0 80 80"
        width="80"
      >
        <defs>
          <clippath
            id="recharts3-clip"
          >
            <rect
              height="70"
              width="70"
              x="5"
              y="5"
            />
          </clippath>
        </defs>
        <g
          class="recharts-layer recharts-pie"
        >
          <g
            class="recharts-layer"
          >
            <g
              class="recharts-layer recharts-pie-sector"
            />
            <g
              class="recharts-layer recharts-pie-sector"
            />
          </g>
        </g>
      </svg>
    </div>
  </div>
  <div
    class="h3 emotion-1"
  >
    30
    %
  </div>
  <div
    class="h6 emotion-2"
  >
    title
  </div>
</div>
`;

exports[`CoverageChart dark 1`] = `
.emotion-3 {
  position: relative;
  width: 120px;
  height: 120px;
}

.emotion-0 {
  position: absolute;
  top: 10px;
  left: 20px;
  width: 80px;
}

.emotion-1 {
  position: absolute;
  top: 10px;
  left: 20px;
  height: 80px;
  width: 80px;
  line-height: 80px;
  text-align: center;
  font-size: 14px;
  font-weight: 400;
}

.emotion-2 {
  position: absolute;
  bottom: 0;
  width: 100%;
  text-align: center;
  padding: 5px 15px;
  margin: 0;
  font-weight: 100;
}

<div
  class=" emotion-3"
>
  <div
    class="emotion-0"
  >
    <div
      class="recharts-wrapper"
      style="position: relative; cursor: default; width: 80px; height: 80px;"
    >
      <svg
        class="recharts-surface"
        height="80"
        version="1.1"
        viewBox="0 0 80 80"
        width="80"
      >
        <defs>
          <clippath
            id="recharts7-clip"
          >
            <rect
              height="70"
              width="70"
              x="5"
              y="5"
            />
          </clippath>
        </defs>
        <g
          class="recharts-layer recharts-pie"
        >
          <g
            class="recharts-layer"
          >
            <g
              class="recharts-layer recharts-pie-sector"
            />
            <g
              class="recharts-layer recharts-pie-sector"
            />
          </g>
        </g>
      </svg>
    </div>
  </div>
  <div
    class="h3 emotion-1"
  >
    50
    %
  </div>
  <div
    class="h6 emotion-2"
  >
    title
  </div>
</div>
`;
